### DISCUSSION TITLE
TypeError: Cannot convert undefined or null to object

### DISCUSSION THREAD

**Discussion by eduardosousaufjf** [quasarframework/quasar#12177] [General - Components / Directives / etc] [answered]
I am trying to use Quasar (Vite Plugin) and Jest but did not succeed. The documentation is not very clear about the process, but I tried to follow it as possible and even used the jest config suggested in the [testing repo](https://github.com/quasarframework/quasar-testing/blob/next/packages/unit-jest/src/templates/typescript/jest.config.js)

I always end up getting a 

` TypeError: Cannot convert undefined or null to object
        at Function.assign (<anonymous>)
 at installQuasar (node_modules/quasar/dist/quasar.cjs.prod.js:6:13567)
      at Object.<anonymous> (src/tests/unit/example.spec.ts:9:30)
`

So I created a repo as small as possible to reproduce this issue:

https://github.com/eduardosousaufjf/quasar-jest

To reproduce the error, please install 

`npm i`

and then run the tests: 

`npm run test:unit:coverage`

Can anyone help me with this?

**Comment by yusufkandemir** [comment, marked as answer]
Probably the most important line in the jest config:
https://github.com/quasarframework/quasar-testing/blob/3904462d492ccafc3887f37879c7a98ad7695269/packages/unit-jest/src/templates/typescript/jest.config.js#L56
```js
    '^quasar$': 'quasar/dist/quasar.esm.prod.js',
```

Using this will surely get rid of the current error. But, I can't guarantee there will not be any other error. I can't run the repro ATM, so I don't know the answer yet.

**Reply by eduardojustos** [reply]
@yusufkandemir if I do so I get a `"SyntaxError: Cannot use import statement outside a module"`

If I use the .cjs file instead of the esm I get a 

`TypeError: Cannot convert undefined or null to object
        at Function.assign (<anonymous>)
`

` at installQuasar (node_modules/quasar/dist/quasar.cjs.prod.js:6:13567)
      at Object.<anonymous> (src/tests/unit/example.spec.ts:9:30)
`

**Reply by yusufkandemir** [reply]
### Adjust babel configuration
https://github.com/quasarframework/quasar-testing/tree/next/packages/unit-jest/src/templates/babelrc

1. Create `/.babelrc`:
```json
{
  "plugins": ["@babel/plugin-syntax-dynamic-import"],
  "env": {
    "test": {
      "plugins": ["dynamic-import-node"],
      "presets": [
        [
          "@babel/preset-env",
          {
            "modules": "commonjs",
            "targets": {
              "node": "current"
            }
          }
        ]
      ]
    }
  }
}
```
2. Add a reference to it in your `babel.config.json`:
```json
{
  "presets": ["@babel/preset-env"],
  "extends": "./.babelrc"
}
```

### Update `jest.config.js`

```diff
--- a/jest.config.js
+++ b/jest.config.js
@@ -47,7 +47,7 @@ module.exports = {
     // See https://github.com/vuejs/vue-jest/issues/188#issuecomment-620750728
     moduleFileExtensions: ['vue', 'js', 'jsx', 'json', 'ts', 'tsx'],
     moduleNameMapper: {
-        'lodash-es': 'lodash',
+        '^quasar$': 'quasar/dist/quasar.esm.prod.js',
         '^~/(.*)$': '<rootDir>/$1',
         '^src/(.*)$': '<rootDir>/src/$1',
         '^app/(.*)$': '<rootDir>/$1',
@@ -71,4 +71,5 @@ module.exports = {
             'jest-transform-stub',
     },
     transformIgnorePatterns: [`node_modules/(?!(${esModules}))`],
+    snapshotSerializers: ['<rootDir>/node_modules/jest-serializer-vue'],
 };
```

You can also remove aliases like `'^boot/(.*)$': '<rootDir>/$1',`, and add the ones from your `tsconfig.json` if you are going to add more of them.

### Update `tsconfig.json`

https://github.com/quasarframework/quasar/blob/15853c8cb123ec196f8c00c36207dcff736b420c/app/tsconfig-preset.json#L3

Add `"allowJs": true,`


### Profit
![image](https://user-images.githubusercontent.com/6266078/151129715-2218e7cd-c959-4563-bcd2-1114dbda7b4e.png)

**Reply by eduardosousaufjf** [reply]
It works! Thank you so much.

**Comment by juanyilxc** [comment]
I also encountered the same problem.     
but it was strange that it was normal on my computer, and it was wrong when was running on gitlab with CI/CD.  
besides it was work before 4 months ago that about Oct 27, 2021.   
I used the `quasar ext add @quasar/testing` command to reinstall the test plugin.     
then I contracted the difference,and found that two main files have changed.    
```diff
modified:   jest.config.js
modified:   package.json
```  
the difference of the package.json file is the depending version in **DevDependencies**   
the difference of the jest.config.js file is as follows.   
```diff
--- a/jest.config.js
+++ b/jest.config.js
@@ -1,4 +1,4 @@
-const esModules = ['quasar/lang', 'lodash-es'].join('|');
+const esModules = ['quasar', 'quasar/lang', 'lodash-es'].join('|');

 /* eslint-env node */
 module.exports = {
@@ -9,6 +9,11 @@ module.exports = {
     'vue-jest': {
       pug: { doctype: 'html' },
     },
+    // Remove if using `const enums`
+    // See https://huafu.github.io/ts-jest/user/config/isolatedModules#example
+    'ts-jest': {
+      isolatedModules: true,
+    },
   },
   // noStackTrace: true,
   // bail: true,
@@ -44,6 +53,7 @@ module.exports = {
   // See https://github.com/vuejs/vue-jest/issues/188#issuecomment-620750728
   moduleFileExtensions: ['vue', 'js', 'jsx', 'json', 'ts', 'tsx'],
   moduleNameMapper: {
+    '^quasar$': 'quasar/dist/quasar.esm.prod.js',
     '^~/(.*)$': '<rootDir>/$1',
     '^src/(.*)$': '<rootDir>/src/$1',
     '^app/(.*)$': '<rootDir>/$1',
```
and I tried 3 times to determine it is `'^quasar$': 'quasar/dist/quasar.esm.prod.js',` and `const esModules = ['quasar', 'quasar/lang', 'lodash-es'].join('|');`   
Although I solved the problem, I don't understand what is caused, and I don't know why this can solve it.  
I hope that people who have understandable can explain it.
